jose-jwt/jwe/JWE.cs

```diff
@@ -0,0 +1,330 @@
+﻿#if NETSTANDARD
+namespace Jose.Jwe
+{
+    using Newtonsoft.Json;
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+
+    public class Recipient
+    {
+        /// <summary
+        /// A recipient for a Jwe
+        /// </summary
+        /// <param name="alg">algorithm to be used to encrypt the CEK (Content Encryption Key).</param>
+        /// <param name="key">key for encrypting CEK (Content Encryption Key). Cannot be null.</param>
+        public Recipient(JweAlgorithm alg, object key, IDictionary<string, object> perRecipientHeaders = null)
+        {
+            this.Alg = alg;
+            this.Key = key ?? throw new ArgumentNullException(nameof(key));
+            this.PerRecipientHeaders = perRecipientHeaders;
+        }
+
+        public JweAlgorithm Alg { get; set; }
+
+        public object Key { get; set; }
+
+        internal IDictionary<string, object> PerRecipientHeaders { get; }
+    }
+
+    public enum SerializationMode
+    {
+        smCompact,
+        smJson,
+    };
+
+    /// <summary>
+    /// Provides methods for encrypting and decrypting using JSON Web Encryption (JWE).
+    /// </summary>
+    public class JWE
+    {
+        /// <summary>
+        /// Encrypts given plaintext using JWE and applies requested encryption/compression algorithms.
+        /// </summary>
+        /// <param name="plaintext">Binary data to encrypt (not null)</param>
+        /// <param name="recipients">The details of who to encrypt the plaintext (or rather the CEK) to.</param>
+        /// <param name="enc">encryption algorithm to be used to encrypt the plaintext.</param>
+        /// <param name="mode">serialization mode to use. Note only one recipient can be specified for compact and flattened json serialization.</param>
+        /// <param name="compression">optional compression type to use.</param>
+        /// <param name="extraHeaders">optional extra headers to put in the JoseProtectedHeader.</param>
+        /// <param name="settings">optional settings to override global DefaultSettings</param>
+        /// <returns>JWT in compact serialization form, encrypted and/or compressed.</returns>
+        public static string Encrypt(byte[] plaintext, IEnumerable<Recipient> recipients, JweEncryption enc, byte[] aad = null, SerializationMode mode = SerializationMode.smCompact, JweCompression? compression = null, IDictionary<string, object> extraHeaders = null, JwtSettings settings = null)
+        {
+            if (plaintext == null)
+            {
+                throw new ArgumentNullException(nameof(plaintext));
+            }
+
+            settings = GetSettings(settings);
+            IJweAlgorithm _enc = settings.Jwe(enc);
+
+            if (_enc == null)
+            {
+                throw new JoseException(string.Format("Unsupported JWE enc requested: {0}", enc));
+            }
+
+            IDictionary<string, object> joseProtectedHeader = MergeHeaders(

```

Comment:
question here, original implementation do not allow to override `enc` and `alg` header values when passing them as `extraHeaders` param. Is it still maintained? Little bit hard to follow logic here as `MergeHeaders()` called multiple times.

Reply:
Yes - I did already add explicit test for `enc `and `alg `overrides - see `Encrypt_WithNonUniqueHeaderParameterNamesInExtraHeaders_Throws`.
That test covers that case explicitly.
There is a slight change in behaviour though (which test documents), that with new implementation the function throws out on attempt to override `enc `or `alg` (or to specify another header twice in both the per-recipient and extra headers), rather than silently ignore a subset of the passed extra parameters.
I'll leave this one open rather than mark resolved in case you have strong opinion on not throwing for 'bad' arguments,

Reply:
Isn't it testing `cty` header ?

Reply:
It is already is checking override for `enc` and `alg`... if you take another look it's an xunit `Theory` test rather than a `Fact`... so it is actually running 3 unit tests separately for the 3 supplied InlineData injectedHeaderName names.  `alg` and `enc` (which can never be overridden), and also for `example.com:extra_recipient_header`, which in the test example has already been set for the recipient so can't be set again in extras. If you try running you'll see the 3 different runs.

```
 [Theory]
 [InlineData("example.com:extra_recipient_header")]
 [InlineData("alg")]
 [InlineData("enc")]

```

Reply:
ok, my bad.